### PR TITLE
Adds context as first-class feature of Model.

### PR DESCRIPTION
**Pull request by mbleigh**
This adds the ability to pass context (an array of `DocumentData[]`) to a model and adds default support for prompt augmentation if a model does not declare specific support for context. Usage:

```ts
const query = "What do cats like to eat?";
const docs = await retrieveStuff({query});

generate({
  prompt: `Answer this question: ${query}`,
  context: docs
});
```

This will augment the prompt to look something like this:

```
Answer this question: What do cats like to eat?

Use the following information to complete your task:

- [0]: Cats eat a variety of foods including dry kibble and wet foods that can be purchased at a pet store.
- [1]: Cats will hunt mice and birds for fun and for food.
```

### CODE REVIEW

File: js/dotprompt/src/prompt.ts [hunk 1]
```diff
@@ -9,7 +9,7 @@
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
- * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+cd * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
```
Comment (maxl0rd) [inline js/dotprompt/src/prompt.ts:12]: oops

**Review by maxl0rd — COMMENTED** [review]
I think this looks good. I like how dotprompt will let you position the context in the template.

Made one comment where I think you could thread it through to vanilla prompts too.

**Review by pavelgj — APPROVED** [review]

File: js/dotprompt/src/template.ts [hunk 2]
```diff
@@ -41,6 +48,23 @@ function mediaHelper(options: Handlebars.HelperOptions) {
...
+//       text += `[${d.metadata[options.hash.cite]}]`;
+//     }
+//     return Promptbars.escapeExpression(text);
+//   });
+
+//   return new Promptbars.SafeString(`<<<dotprompt:section context>>>
+// - ${items.join('\n- ')}`);
+// }
+// Promptbars.registerHelper('context', contextHelper);
+
```
Comment (ssbushi) [inline js/dotprompt/src/template.ts:67]: Remove?

File: js/dotprompt/src/template.ts [hunk 3]
```diff
@@ -103,16 +134,19 @@ export function compile<Variables = any>(
       media: true,
       role: true,
       history: true,
+      // context: true,
```
Comment (ssbushi) [inline js/dotprompt/src/template.ts:137]: Remove.

File: js/dotprompt/tests/template_test.ts [hunk 4]
```diff
@@ -106,6 +106,69 @@ describe('compile', () => {
       template: '{{json . indent=2}}',
       want: [{ role: 'user', content: [{ text: '{\n  "test": true\n}' }] }],
     },
+    // {
+    //   should: 'allow defining context',
```
Comment (ssbushi) [inline js/dotprompt/tests/template_test.ts:110]: Remove?

**Review by ssbushi — APPROVED** [review]

**Pull request merged**